concordia/language_model/local_models.py

```diff
@@ -0,0 +1,307 @@
+                
+                if self._logits:
+                    outputs = self._client(**inputs)
+                    response = outputs[-1]
+                else:
+                    outputs = self._client.generate(**inputs, max_new_tokens=max_tokens, temperature=temperature, seed=seed)
+                    response = self._tokenizer.decode(outputs[0], skip_special_tokens=True)
+
+            except ValueError as e:
+                self.logger.error(f"Error while calling LLM with input: {prompt}. Attempt {retry+1} failed.")
+                if retry == _MAX_SAMPLE_TEXT_ATTEMPTS - 1: 
+                    self.logger.error(f"Max retries exceeded. Raising exception.")
+                    raise language_model.InvalidResponseError(prompt)
+            else:
+                self.logger.debug(f"Succeeded after {retry+1} attempts.")
+                self.logger.debug(f"Response from LLM: {response}")
+                break
+
+        if self._measurements is not None:
+            self._measurements.publish_datum(
+                self._channel,
+                {'raw_text_length': len(response)},
+            )
+        return response
+
+    @staticmethod
+    def extract_choices(text):
+        match = re.search(r'\(?(\w)\)', text)
+        if match:
+            return match.group(1)
+        else:
+            return None
+
+    @override
+    def sample_choice(
+        self,
+        prompt: str,
+        responses: Sequence[str],
+        *,
+        seed: int | None = None,
+    ) -> tuple[int, str, dict[str, float]]:
+        max_characters = len(max(responses, key=len))
+
+        attempts = 1
+        for _ in range(_MAX_MULTIPLE_CHOICE_ATTEMPTS):
+            sample = self.sample_text(
+                prompt,
+                max_characters=max_characters,
+                temperature=0.0,
+                seed=seed,
+                system_message=MULTIPLE_CHOICE_SYSTEM_MESSAGE,
+            )
+            answer = self.extract_choices(sample)
+            try:
+                idx = responses.index(answer)
+            except ValueError:
+                attempts += 1
+                continue
+            else:
+                if self._measurements is not None:
+                    self._measurements.publish_datum(
+                        self._channel, {'choices_calls': attempts}
+                    )
+                debug = {}
+                return idx, responses[idx], debug
+
+        self.logger.error(f"Multiple choice failed after {_MAX_MULTIPLE_CHOICE_ATTEMPTS} attempts.\nLLM Input: {prompt}\nLLM Output: {sample}\nExtracted Answer: {answer}")
+
+        raise language_model.InvalidResponseError(
+             f'Too many multiple choice attempts.\nLLM Input: {prompt}\nLLM Output: {sample}'
+        )
+
+# Setup sentence encoder
+from sentence_transformers import SentenceTransformer
+def embedding_model(model: str = 'sentence-transformers/sentence-t5-base') -> SentenceTransformer:
```

Comment:
I don't think this is used anywhere and it does very little, I suggest omitting it.

Reply:
done